### PR TITLE
DeviceManager dropdown only appears if there are profiles for the service.

### PR DESCRIPTION
**Pull request by RogPodge**
## Overview
Made it so that the user is only prompted to specify a DeviceManagerProfile if there are valid ones to choose from, otherwise, it is hidden from the user.

This addresses the clutter and confusion that could occur with first time users stumbling through sections like the InputDataProviders.

old:
image
image

new:
image

### CODE REVIEW

**Comment by keveleigh** [comment]
@RogPodge Looks like you caught the dev branch in a broken state. #9416 will fix it

**Comment by RogPodge** [comment]
Now refactored to support the scenario where there are multiple profile types encoded for a single device manager (even though this case should be rare and doesn't currently exist in the code from what I can see)

DeviceManagerMultitype

**Comment by RogPodge** [comment]
image
 
Users can now no longer accidentally select (None) for DeviceManagers which require profiles

File: Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs [hunk 1]
```diff
@@ -531,27 +531,44 @@ public static void DrawSubProfileEditor(Object profileObject, bool renderProfile
             EditorGUILayout.EndVertical();
         }
 
+        public static bool DrawProfileDropDownList(SerializedProperty property, BaseMixedRealityProfile profile, Object oldProfileObject, Type profileType, bool requiresProfile, bool showAddButton)
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs:534]: nit: missing summary

File: Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs [hunk 2]
```diff
@@ -531,27 +531,44 @@ public static void DrawSubProfileEditor(Object profileObject, bool renderProfile
...
+        {
+            return DrawProfileDropDownList(property, profile, oldProfileObject, new Type[] { profileType }, requiresProfile, showAddButton);
+        }
+
         /// <summary>
         /// Draws a dropdown with all available profiles of profilyType.
         /// </summary>
         /// <returns>True if property was changed.</returns>
-        public static bool DrawProfileDropDownList(SerializedProperty property, BaseMixedRealityProfile profile, Object oldProfileObject, Type profileType, bool showAddButton)
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs:538]: Does this old method signature still exist? We should re-add it if not and mark it as obsolete

File: Assets/MRTK/Core/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs [hunk 3]
```diff
@@ -201,5 +202,29 @@ protected static bool IsProfileLock(BaseMixedRealityProfile profile)
...
+        /// True if the service is decorated with an attribute indicating a profile is required, false otherwise.
+        /// </returns>
+        protected static bool IsProfileRequired(SystemType serviceType)
+        {
+            // Services marked with the MixedRealityExtensionServiceAttribute (or a derivative)
+            // support specifying whether or not a profile is required.
+            return IsProfileRequired(serviceType?.Type);
+        }
+
+        protected static bool IsProfileRequired(Type serviceType)
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs:220]: missing summary

File: Assets/MRTK/Core/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs [hunk 4]
```diff
@@ -201,5 +202,29 @@ protected static bool IsProfileLock(BaseMixedRealityProfile profile)
...
+        {
+            if(serviceType != null)
+            {
+                MixedRealityExtensionServiceAttribute attribute = (serviceType != null) ? MixedRealityExtensionServiceAttribute.Find(serviceType) : null;
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs:224]: Looks like we're null checking `serviceType` twice here. Could this be simplified a bit?

File: Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs (hunk 2 above)
Comment (RogPodge) [reply Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs:538]: Yeah, the function still signature still exists, it's a wrapper for this function now. (line 534)

**Review by keveleigh — APPROVED** [review]

File: Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs [hunk 5]
```diff
@@ -562,7 +583,8 @@ public static bool DrawProfileDropDownList(SerializedProperty property, BaseMixe
                     profileContent,
                     GUILayout.ExpandWidth(true));
 
-                property.objectReferenceValue = (newIndex > 0) ? profileInstances[newIndex - 1] : null;
+                int profileInstanceIndex = newIndex - dropdownOffset;
+                property.objectReferenceValue = (profileInstanceIndex >= 0) ? profileInstances[profileInstanceIndex] : null;
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs:587]: How does this behave if no profile assets are found? Does it just say `null` or something in the dropdown?

Comment (RogPodge) [reply Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs:587]: It breaks currently (yikes!) Making a fix for this case now

Comment (RogPodge) [reply Assets/MRTK/Core/Inspectors/Utilities/MixedRealityInspectorUtility.cs:587]: Addressed this case and added a warning message, let me know if it should be worded to be clearer!

image

**Comment by RogPodge** [comment]
Last set of modifications to be performant on Unity 2019+. In Unity 2018, there is a noticeable delay due to cycling through all of the types in the assembly, so we're having it use the old behavior for performance reasons (and the edge cases where it fails are few and far between)

File: Assets/MRTK/Core/Inspectors/Utilities/MixedRealityProfileUtility.cs [hunk 6]
```diff
@@ -130,6 +129,18 @@ public static IReadOnlyCollection<Type> GetProfileTypesForService(Type serviceTy
             if (!profileTypesForServiceCaches.TryGetValue(serviceType, out types))
             {
                 HashSet<Type> allTypes = new HashSet<Type>();
+
+#if UNITY_2019_1_OR_NEWER
+                foreach (var type in TypeCache.GetTypesDerivedFrom<BaseMixedRealityProfile>())
+                {
+                    if (IsProfileForService(type, serviceType))
+                    {
+                        {
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Utilities/MixedRealityProfileUtility.cs:138]: nit: extra braces

File: Assets/MRTK/Core/Inspectors/Utilities/MixedRealityProfileUtility.cs [hunk 7]
```diff
@@ -130,6 +129,18 @@ public static IReadOnlyCollection<Type> GetProfileTypesForService(Type serviceTy
...
+                            allTypes.Add(type);
+                        }
+                    }
```
Comment (keveleigh) [inline Assets/MRTK/Core/Inspectors/Utilities/MixedRealityProfileUtility.cs:141]:
```suggestion
                    {
                        allTypes.Add(type);
                    }
```

**Pull request merged**